cli/lsp/sources.rs

```diff
@@ -40,16 +40,108 @@ pub async fn cache(
   builder.add(specifier, false).await
 }
 
+fn get_remote_headers(
+  cache_filename: &Path,
+) -> Option<HashMap<String, String>> {
+  let metadata_path = http_cache::Metadata::filename(cache_filename);
+  let metadata_str = fs::read_to_string(metadata_path).ok()?;
+  let metadata: http_cache::Metadata =
+    serde_json::from_str(&metadata_str).ok()?;
+  Some(metadata.headers)
+}
+
+fn resolve_remote_specifier(
+  specifier: &ModuleSpecifier,
+  http_cache: &HttpCache,
+  redirect_limit: isize,
+) -> Option<ModuleSpecifier> {
+  let cache_filename = http_cache.get_cache_filename(specifier.as_url());
+  if redirect_limit >= 0 && cache_filename.is_file() {
+    let headers = get_remote_headers(&cache_filename)?;
+    if let Some(location) = headers.get("location") {
+      let redirect =
+        ModuleSpecifier::resolve_import(location, specifier.as_str()).ok()?;
+      resolve_remote_specifier(&redirect, http_cache, redirect_limit - 1)
```

Comment:
What happens when my malicious server sends back a `Location: file:///etc/passwd` response? It'll follow the redirect, wouldn't it?

It looks like `cli/file_fetcher.rs` is susceptible to that, too.

Reply:
Remote modules are banned from importing local modules. It should happen at another point of module resolution but I am not sure we test for it.

Reply:
Ok, I investigated this a bit.  It doesn't specifically fail in the way I expected it to fail, but it fails.  When it tries to follow it, and it does `http_cache.get_cache_filename(specifier.as_url())` it panics with an unimplemented in `http_cache::base_url_to_filename`, as it tries to lookup the cache for a local file.

I think `cli/file_fetcher.rs` would have a slightly different behaviour.  It would allow the module to be resolved, but in the module graph, it would eventually surface up a `GraphError::InvalidLocalImport` when that module is attempted to be loaded.

I'm going to look into how difficult it would be to not throw in `get_cache_filename` as it would be better for the lsp to simply resolve to a `None` instead of panicing.

Reply:
Ok, I changed it so it doesn't panic, and trying to redirect effectively results in the module being `None`.